File: android/nativebrik/src/main/java/com/nativebrik/sdk/template/formatter.kt
package com.nativebrik.sdk.template

import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.jsonPrimitive

private fun defaultFormatter(value: JsonElement?): String {
    if (value == null) {
        return ""
    }
    if (value is JsonNull) {
        return ""
    }
    if (value is JsonPrimitive) {
        return value.jsonPrimitive.content
    }
    return value.toString()
}

private fun uppercaseFormatter(value: JsonElement?): String {
    return defaultFormatter(value).uppercase()
}
private fun lowercaseFormatter(value: JsonElement?): String {
    return defaultFormatter(value).lowercase()
}

private fun jsonFormatter(value: JsonElement?): String {
    return Json.encodeToString(value)
}

internal fun formatValue(formatter: String, value: JsonElement?): String {
    return when (formatter) {
        "json" -> jsonFormatter(value)
        "upper" -> uppercaseFormatter(value)
        "lower" -> lowercaseFormatter(value)
        else -> defaultFormatter(value)
    }
}

File: Sources/Nativebrik/template/formatter.swift
//
//  formatter.swift
//  Nativebrik
//
//  Created by Ryosuke Suzuki on 2024/03/07.
//

import Foundation

fileprivate func defaultFormatter(_ value: Any?) -> String {
    return value.flatMap({ String.init(describing: $0 )}) ?? ""
}

fileprivate func jsonFormatter(_ value: Any?) -> String {
    do {
        guard let value = value else {
            return "null"
        }
        if (JSONSerialization.isValidJSONObject(value)) {
            let jsonData = try JSONSerialization.data(withJSONObject: value)
            return String(decoding: jsonData, as: UTF8.self)
        } else {
            let jsonData = try JSONSerialization.data(withJSONObject: [value])
            let jsonStr = String(decoding: jsonData, as: UTF8.self)
            let trimmed = jsonStr.dropFirst().dropLast().trimmingCharacters(in: .whitespaces)
            return trimmed
        }
    } catch {
        return "null"
    }
}
fileprivate func uppercaseFormatter(_ value: Any?) -> String {
    let str = defaultFormatter(value)
    return str.uppercased()
}

fileprivate func lowercaseFormatter(_ value: Any?) -> String {
    let str = defaultFormatter(value)
    return str.lowercased()
}

func formatValue(formatter: String, value: Any?) -> String {
    switch (formatter) {
    case "json":
        return jsonFormatter(value)
    case "upper":
        return uppercaseFormatter(value)
    case "lower":
        return lowercaseFormatter(value)
    default:
        return defaultFormatter(value)
    }
}
